### PR TITLE
improve validation of llm_config

### LINKED ISSUE #1863: Autogen llm_config error
Hi, when my version is `pyautogen==0.2.16` Autogen gives this error:

```terminal
ValueError: Please either set llm_config to False, or specify a non-empty 'model' either in 'llm_config' or in each config of 'config_list'.
``` 

But when i switch to the `pyautogen==0.2.2` version. It does not gives any error.

This is my llm_config:
```py
{'config_list': 
        [{
        'base_url': 'http://0.0.0.0:8053',
        'api_key': 'NULL'
    }],
    'cache_seed': 25}
```

And litellm runs on that port without any issues.

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

The current llm_config validation is too restricted and excludes valid configs like #1863

## Related issue number

close #1863

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by sonichi** [comment]
@doganaktar could you test whether this PR solves your issue?

**Review by davorrunje — APPROVED** [review]

**Comment by davorrunje** [comment]
@sonichi I fixed `test_register_for_llm_without_LLM()`

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -139,20 +140,14 @@ def __init__(
             else (lambda x: content_str(x.get("content")) == "TERMINATE")
         )
 
-        if llm_config is False:
+        if llm_config is False or llm_config is None and self.DEFAULT_CONFIG is None:
```
Comment (jackgerrits) [inline autogen/agentchat/conversable_agent.py:143]: When would DEFAULT_CONFIG not be None here?

File: autogen/agentchat/user_proxy_agent.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def __init__(
                 - timeout (Optional, int): The maximum execution time in seconds.
                 - last_n_messages (Experimental, Optional, int): The number of messages to look back for code execution. Default to 1.
             default_auto_reply (str or dict or None): the default auto reply message when no code execution or llm based reply is generated.
-            llm_config (dict or False): llm inference configuration.
+            llm_config (dict or False or None): llm inference configuration.
```
Comment (jackgerrits) [inline autogen/agentchat/user_proxy_agent.py:73]: What is the difference between passing False and None?

Comment (sonichi) [reply autogen/agentchat/user_proxy_agent.py:73]: False: disable LLM
None: using self.DEFAULT_CONFIG
Does this make sense? I can add this to the docstr if it makes sense.

File: test/agentchat/test_conversable_agent.py [hunk 3]
```diff
@@ -817,25 +817,29 @@ def exec_python(cell: Annotated[str, "Valid Python cell to execute."]) -> str:
```
Comment (olgavrou) [inline test/agentchat/test_conversable_agent.py:818]: should more test cases be added here where the default dict is not set to None?


Also, custom models don't need the `model` field and will fail unnecessarily if it is missing (another potential test case to be covered)

**Comment by sonichi** [comment]
> @sonichi I fixed `test_register_for_llm_without_LLM()`

Thanks!

File: test/agentchat/test_conversable_agent.py (hunk 3 above)
Comment (sonichi) [reply test/agentchat/test_conversable_agent.py:818]: Could you add one? OK to add in a separate PR.

Comment (olgavrou) [reply test/agentchat/test_conversable_agent.py:818]: won't it fail since the requirement is still that model is required?

Comment (sonichi) [reply test/agentchat/test_conversable_agent.py:818]: That requirement is removed in this PR.

**Comment by jackgerrits** [comment]
Meta point here - this issue is an argument I would use for requiring the user to actually pass in the `ModelClient` they want to use into the constructor of `ConversableAgent`. We're trying to make a generic configuration scheme fit models with varying configuration requirements. 

What if instead it looked like:
```python
from autogen import AssistantAgent, DefaultModelClientFactory
import os

llm_config = {
    "config_list": [{"model": "gpt-4", "api_key": os.environ["OPENAI_API_KEY"]}],
}
model_client = DefaultModelClientFactory.create_from_config(llm_config)

assistant = AssistantAgent("assistant", model_client=model_client)
```

File: autogen/agentchat/user_proxy_agent.py (hunk 2 above)
Comment (jackgerrits) [reply autogen/agentchat/user_proxy_agent.py:73]:
```
llm_config: Optional[Union[Dict, Literal[False]]] = ConversableAgent.DEFAULT_CONFIG,
```

File: autogen/agentchat/conversable_agent.py (hunk 1 above)
Comment (jackgerrits) [reply autogen/agentchat/conversable_agent.py:144]: Is DEFAULT_CONFIG class or instance scoped? Either they'd have to subclass it or change it for all usages of the class. It doesn't feel very easy to understand

**Comment by olgavrou** [comment]
> Meta point here - this issue is an argument I would use for requiring the user to actually pass in the `ModelClient` they want to use into the constructor of `ConversableAgent`. We're trying to make a generic configuration scheme fit models with varying configuration requirements.
> 
> What if instead it looked like:
> 
> ```python
> from autogen import AssistantAgent, DefaultModelClientFactory
> import os
> 
> llm_config = {
>     "config_list": [{"model": "gpt-4", "api_key": os.environ["OPENAI_API_KEY"]}],
> }
> model_client = DefaultModelClientFactory.create_from_config(llm_config)
> 
> assistant = AssistantAgent("assistant", model_client=model_client)
> ```

I am not opposed to having the model client passed in the constructor, I proposed this during the original design of the custom model feature and happy to have that be the case

File: autogen/agentchat/user_proxy_agent.py (hunk 2 above)
Comment (sonichi) [reply autogen/agentchat/user_proxy_agent.py:73]: I guess a sentinel is better. I'm not used to it yet. I'm using None as a Sentinel.

**Comment by sonichi** [comment]
> Meta point here - this issue is an argument I would use for requiring the user to actually pass in the `ModelClient` they want to use into the constructor of `ConversableAgent`. We're trying to make a generic configuration scheme fit models with varying configuration requirements.
> 
> What if instead it looked like:
> 
> ```python
> from autogen import AssistantAgent, DefaultModelClientFactory
> import os
> 
> llm_config = {
>     "config_list": [{"model": "gpt-4", "api_key": os.environ["OPENAI_API_KEY"]}],
> }
> model_client = DefaultModelClientFactory.create_from_config(llm_config)
> 
> assistant = AssistantAgent("assistant", model_client=model_client)
> ```

I'd like to keep user's code simple for now.

File: autogen/agentchat/user_proxy_agent.py (hunk 2 above)
Comment (davorrunje) [reply autogen/agentchat/user_proxy_agent.py:73]: `None` is the preferred sentinel value unless it is a valid parameter value. If such cases, `object()` is used instead.

Comment (davorrunje) [reply autogen/agentchat/user_proxy_agent.py:73]: yes, this is confusing.

Comment (sonichi) [reply autogen/agentchat/user_proxy_agent.py:73]: I gave it another try to make None as a sentinel. None and False have different meanings: False means disabling llm, and None means using `self.DEFAULT_CONFIG`.

**Comment by sonichi** [comment]
Merging for unblocking the release that many users demanded.
